xarray/core/indexing.py

```diff
@@ -651,12 +676,14 @@ def __init__(self, array):
         self.array = _wrap_numpy_scalars(as_indexable(array))
 
     def _ensure_cached(self):
-        if not isinstance(self.array, NumpyIndexingAdapter):
-            self.array = NumpyIndexingAdapter(np.asarray(self.array))
+        self.array = as_indexable(self.array.get_duck_array())
```

Comment:
I think this is the right generalization. `as_indexable` should choose an appropriate indexing adapter.